### ISSUE TITLE
Enhancement: Add support for filtering files in Globbing alias

### ISSUE THREAD

**Issue by kcamp** [cake-build/cake#1384]
`GlobbingAliases.cs` has support for filtering the globbing traversal at the directory level, but no support for filtering files.  I have found that support for logic that is more complex than the globbing expression allows has been useful, and have seen gitter traffic for the same issue.

This would be a trivial addition to make at the method alias level.  I would propose that two overloads be created --

```cs
GetFiles(this ICakeContext context, string pattern, Func<IFile, bool> filePredicate)
 ...

GetFiles(this ICakeContext context, string pattern, Func<IDirectory, bool> predicate, Func<IFile, bool> filePredicate)
...
```

The former would satisfy the direct need of the feature request, the latter would provide a graceful pairing of current + extended logic.

Assuming that this is something that seems useful and aligned with the intent of the project, the question I would raise is whether it's more appropriate to implement the logic directly in the alias method, similar to .. 

```
// glob the files, and then filter down based on the filePredicate
var files = context.Globber.Match(pattern).OfType<FilePath>()
                            .Where(file => filePredicate(context.FileSystem.GetFile(file)));
```

or whether it would be more appropriately implemented in the globbing internals -- `GlobVisitor`, `GlobVisitorContext`, and so on. 

Constraining it to the alias would certainly be less involved, but the latter seems like it would probably be a cleaner (and faster) implementation, particularly in eliminating the extra calls to `FileSystem.GetFile(...)` -  it seems quite apt that it would amount to refactoring the `GlobVisitor` and `GlobVisitorContext` so that the logic in `FindCandidates` would be extended to encompass something like .. 

```
if (includeFiles)
{
    foreach (var file in current.GetFiles("*", option))
    {
        var lastPath = file.Path.Segments.Last();
        if (node.IsMatch(lastPath) && context.ShouldInclude(file))
        ....
```

Thoughts on whether or not this would be worth pursuing?  Any other feedback?

**Comment by kcamp** [comment]
For what it's worth & work doesn't get duplicated, I've made good progress in extending the globbing to support this and expect to have a PR for it in the next couple of days.

**Comment by devlead** [comment]
Fixed by #2458
